packages/auth/src/AuthProvider.tsx

```diff
@@ -190,171 +274,169 @@ export class AuthProvider extends React.Component<
-    await this.rwClient.logout(options)
-    this.setState({
-      userMetadata: null,
-      currentUser: null,
-      isAuthenticated: false,
-      hasError: false,
-      error: undefined,
-    })
-  }
+  /** Whenever the rwClient is ready to go, restore auth and reauthenticate */
+  useEffect(() => {
+    if (rwClient && !hasRestoredState) {
+      setHasRestoredState(true)
 
-  signUp = async (options?: any) => {
-    const signupOutput = await this.rwClient.signup(options)
-    await this.reauthenticate()
-    return signupOutput
-  }
+      const doRestoreState = async () => {
+        await rwClient.restoreAuthState?.()
+        reauthenticate()
+      }
 
-  forgotPassword = async (username: string) => {
-    if (this.rwClient.forgotPassword) {
-      return await this.rwClient.forgotPassword(username)
-    } else {
-      throw new Error(
-        `Auth client ${this.rwClient.type} does not implement this function`
-      )
+      doRestoreState()
     }
-  }
+  }, [rwClient, reauthenticate, hasRestoredState])
 
-  resetPassword = async (options?: any) => {
-    if (this.rwClient.resetPassword) {
-      return await this.rwClient.resetPassword(options)
-    } else {
-      throw new Error(
-        `Auth client ${this.rwClient.type} does not implement this function`
-      )
-    }
-  }
+  const { client, type, children } = props
 
-  validateResetToken = async (resetToken: string | null) => {
-    if (this.rwClient.validateResetToken) {
-      return await this.rwClient.validateResetToken(resetToken)
-    } else {
-      throw new Error(
-        `Auth client ${this.rwClient.type} does not implement this function`
-      )
-    }
-  }
-
-  render() {
-    const { client, type, children } = this.props
-
-    return (
-      <AuthContext.Provider
-        value={{
-          ...this.state,
-          logIn: this.logIn,
-          logOut: this.logOut,
-          signUp: this.signUp,
-          getToken: this.getToken,
-          getCurrentUser: this.getCurrentUser,
-          hasRole: this.hasRole,
-          reauthenticate: this.reauthenticate,
-          forgotPassword: this.forgotPassword,
-          resetPassword: this.resetPassword,
-          validateResetToken: this.validateResetToken,
-          client,
-          type: type as SupportedAuthTypes,
-        }}
-      >
-        {children}
-      </AuthContext.Provider>
-    )
-  }
+  return (
+    <AuthContext.Provider
+      value={{
+        ...authProviderState,
+        logIn,
+        logOut,
+        signUp,
+        getToken,
+        getCurrentUser,
+        hasRole,
+        reauthenticate,
+        forgotPassword,
+        resetPassword,
+        validateResetToken,
+        client,
+        type: type as SupportedAuthTypes,
+      }}
+    >
+      {children}
+      {rwClient?.useListenForUpdates !== undefined && (
+        <AuthUpdateListener
+          // NOTE: Typescript doesn't recognize that we've just validated that this client is defined
```

Comment:
What if you made a custom type guard here, like `isAuthClientWithListener`*? Could you get rid of the `as any` cast then?

*) Feel free to come up with a better name

Reply:
AHA perfect - didn't think to use a type guard function here but that worked perfectly.